### PR TITLE
LLL: introduce testing framework

### PR DESCRIPTION
**Pull request by axic**
This splits the generic part of ExecutionFramework from Solidity specific code.

I think after this we could look at merging ExecutionFramework and RPCSession or parts thereof?

### CODE REVIEW

File: liblll/Compiler.cpp [hunk 1]
```diff
@@ -99,14 +99,15 @@ std::string dev::eth::compileLLLToAsm(std::string const& _src, bool _opt, std::v
...
 	try
 	{
+		sp::utree o;
 		parseTreeLLL(_src, o);
+		ostringstream ret;
+		debugOutAST(ret, o);
+		killBigints(o);
+		return ret.str();
 	}
 	catch (...) {}
```
Comment (chriseth) [inline liblll/Compiler.cpp]: This is really horrible. Really, really horrible.

File: test/CMakeLists.txt [hunk 2]
```diff
@@ -29,6 +30,7 @@ set(EXECUTABLE soltest)
 eth_simple_add_executable(${EXECUTABLE} ${SRC_LIST} ${HEADERS})
 
 eth_use(${EXECUTABLE} REQUIRED Solidity::solidity)
+eth_use(${EXECUTABLE} REQUIRED Solidity::lll)
```
Comment (chriseth) [inline test/CMakeLists.txt]: I think you can combine that with the previous line.

File: test/ExecutionFramework.h [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+	}
+
+	template <class... Args>
+	bytes const& callContractFunction(std::string _sig, Args const&... _arguments)
+	{
+		return callContractFunctionWithValue(_sig, 0, _arguments...);
+	}
+
+	template <class CppFunction, class... Args>
+	void testSolidityAgainstCpp(std::string _sig, CppFunction const& _cppFunction, Args const&... _arguments)
```
Comment (chriseth) [inline test/ExecutionFramework.h]: Could be renamed.

File: test/ExecutionFramework.h [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+			solidityResult == cppResult,
+			"Computed values do not match.\nSolidity: " +
+				toHex(solidityResult) +
+				"\nC++:      " +
+				toHex(cppResult)
+		);
+	}
+
+	template <class CppFunction, class... Args>
+	void testSolidityAgainstCppOnRange(std::string _sig, CppFunction const& _cppFunction, u256 const& _rangeStart, u256 const& _rangeEnd)
```
Comment (chriseth) [inline test/ExecutionFramework.h]: Could be renamed.

File: liblll/Compiler.cpp (hunk 1 above)
Comment (axic) [reply liblll/Compiler.cpp]: It is, but at the same time the rest of the class must be fixed as they are all broken in the same way.

**Comment by axic** [comment]
@chriseth updated with the changes requested

File: liblll/Compiler.cpp [hunk 5]
```diff
@@ -99,14 +99,28 @@ std::string dev::eth::compileLLLToAsm(std::string const& _src, bool _opt, std::v
...
 	sp::utree o;
+
 	try
 	{
 		parseTreeLLL(_src, o);
 	}
-	catch (...) {}
+	catch (...)
+	{
+		failed = true;
```
Comment (chriseth) [inline liblll/Compiler.cpp]: We can keep it like that, but in this case, I would prefer to use a `return string()` here.

Comment (axic) [reply liblll/Compiler.cpp]: The `killBigints(o);` part needs to be executed even in that case to potentially deallocate data.

Comment (axic) [reply liblll/Compiler.cpp]: Changed.

**Review by chriseth — APPROVED** [review]

**Pull request merged**